pkg/apis/sources/v1alpha1/rabbitmq_validation.go

```diff
@@ -19,40 +19,37 @@ package v1alpha1
 import (
 	"context"
 
-	"github.com/google/go-cmp/cmp/cmpopts"
-
 	"knative.dev/pkg/apis"
 	"knative.dev/pkg/kmp"
 )
 
 func (current *RabbitmqSource) Validate(ctx context.Context) *apis.FieldError {
-	if !current.Spec.Predeclared && current.Spec.ExchangeConfig.Name == "" {
-		return &apis.FieldError{
-			Message: "Name of exchange must be provided when spec.predeclared is false",
-			Paths:   []string{"spec", "exchangeConfig", "name"},
+	if current.Spec.RabbitmqResourcesConfig == nil {
+		return apis.ErrMissingField("rabbitmqResourcesConfig").ViaField("spec")
+	} else {
+		if !current.Spec.RabbitmqResourcesConfig.Predeclared && current.Spec.RabbitmqResourcesConfig.ExchangeName == "" {
+			return &apis.FieldError{
+				Message: "Name of exchange must be provided when spec.predeclared is false",
+				Paths:   []string{"spec", "rabbitmqResourcesConfig", "name"},
+			}
 		}
 	}
 
-	var errs *apis.FieldError
 	if current.Spec.RabbitmqClusterReference == nil {
 		return apis.ErrMissingField("rabbitmqClusterReference").ViaField("spec")
 	} else {
 		if current.Spec.RabbitmqClusterReference.Name == "" {
-			return errs.Also(apis.ErrMissingField("name").ViaField("rabbitmqClusterReference").ViaField("spec"))
-		}
-		if current.Spec.RabbitmqClusterReference.Namespace == "" {
-			return errs.Also(apis.ErrMissingField("namespace").ViaField("rabbitmqClusterReference").ViaField("spec"))
-		}
-		if current.Spec.RabbitmqClusterReference.ConnectionSecret != nil {
+			if current.Spec.RabbitmqClusterReference.ConnectionSecret == nil {
+				return apis.ErrMissingField("name").ViaField("rabbitmqClusterReference").ViaField("spec")
+			}
+		} else if current.Spec.RabbitmqClusterReference.ConnectionSecret != nil {
 			return apis.ErrDisallowedFields("connectionSecret").ViaField("rabbitmqClusterReference").ViaField("spec")
```

Comment:
It'll be worthwhile having a more descriptive error here rather than `connectionSecret` is `Disallowed`. Something like both name+namespace and connection secret cannot be set.

Reply:
Ok, but since there is a refactor comming regarding this validation, can this be moved to the next pr?